[WIP] Add python and markdown pre-commit hooks

zhiltsov-max opened the pull request:
<!-- Raised an issue to propose your change (https://github.com/cvat-ai/cvat/issues).
It helps to avoid duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [CONTRIBUTION](https://github.com/cvat-ai/cvat/blob/develop/CONTRIBUTING.md)
guide. -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
<!-- Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here. Describe your changes in detail, add
screenshots. -->

- Added `pylint`, `black` and `isort` to precommit hooks for updated python files
- Added `remark` checks for markdown files

### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable by a reason then ~~explicitly strikethrough~~ the whole
line. If you don't do that github will show an incorrect process for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I submit my changes into the `develop` branch
- [ ] I have added a description of my changes into [CHANGELOG](https://github.com/cvat-ai/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/cvat-ai/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/cvat-ai/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/cvat-ai/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/cvat-ai/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/cvat-ai/cvat/tree/develop/cvat-ui#versioning))

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/cvat-ai/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

nmanovic commented:
@SpecLad , could you please review?

SpecLad commented on site/content/en/docs/contributing/running-tests.md:
````diff
@@ -44,6 +44,9 @@ yarn run cypress:run:chrome:canvas3d
    ```
 
+   Please check the [SDK developer guide](/docs/api_sdk/sdk/developer-guide)
+   to install it for local development and to run tests.
````
What is "it"?

SpecLad commented on site/content/en/docs/contributing/coding-style.md:
```diff
@@ -5,5 +5,29 @@ weight: 4
+## Hooks
+
+It is recommended to use [git pre-commit hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks)
+to avoid the need of manual control of linting and formatting the code. We provide a set of hooks
```
Isn't it just one hook (`pre-commit`)?

zhiltsov-max replied:
Maybe, what is the problem with this? This one already includes a bunch of calls, and I clearly see it can be extended. I tried writing some custom scripts in the commits above, it turned out to be too complicated to maintain manually. Having everything handled by `list-staged` greatly helps.

SpecLad replied:
I'm just talking about the wording here. We only provide one hook (which does multiple things).

zhiltsov-max commented on lint-staged.config.js:
```diff
@@ -13,28 +13,49 @@ function makePattern(extension) {
+    // Otherwise, they all will use the same config. This will mess up
+    // the current package detection for different modules, leading
+    // to incorrect formatting of module imports.
+    tasks.push(['black', cvatSdk.join(' ')]);
```
1. I can add it, but the file paths are already absolute. It's unlikely that they can be recognized as options
2. What are your typical actions after you know there are formatting problems? I myself usually launch a spell from the bash history, which looks this way now: `black cvat-sdk && isort cvat-sdk && black cvat-cli && isort cvat-cli && black tests/python/ && isort tests/python/`. And this is what I would like to eliminate from user workflows. With autoformatting there is just less extra things to keep in mind, which is good. No extra command from user saves time.

SpecLad replied:
> With autoformatting there is just less extra things to keep in mind, which is good. No extra command from user saves time.

Okay, let's give it a try. As long as all files are already formatted when this PR is merged, it should be fine.

zhiltsov-max commented on lint-staged.config.js:
```diff
@@ -13,28 +13,49 @@ function makePattern(extension) {
+    tasks.push(['isort', cvatSdk.join(' ')]);
+    tasks.push(['isort', cvatCli.join(' ')]);
+    tasks.push(['isort', pythonTests.join(' ')]);
+    tasks.push(['pylint', pythonFiles.join(' ')]);
```
I need to look into this. This is the same that CI checks, probably some files are still not updated since pylint introduction.

zhiltsov-max replied:
I added some files to be ignored by pylint. Migrations are not covered by this, so they will produce warnings when the files are touched. It must be fixed in another PR, but it goes out of the scope for this PR. The problematic migration code itself is, actually, broken, but its dangerous to edit it, because fixes can damage DB.

SpecLad replied:
I still see some errors even without counting migrations:

```
************* Module engine.views
cvat/apps/engine/views.py:309:4: W0221: Variadics removed in overridden 'ProjectViewSet.perform_create' method (arguments-differ)
cvat/apps/engine/views.py:817:4: W0221: Variadics removed in overridden 'TaskViewSet.perform_create' method (arguments-differ)
cvat/apps/engine/views.py:1744:4: W0221: Variadics removed in overridden 'IssueViewSet.perform_create' method (arguments-differ)
cvat/apps/engine/views.py:1819:4: W0221: Variadics removed in overridden 'CommentViewSet.perform_create' method (arguments-differ)
************* Module organizations.views
cvat/apps/organizations/views.py:204:4: W0221: Variadics removed in overridden 'InvitationViewSet.perform_create' method (arguments-differ)
************* Module script
utils/dicom_converter/script.py:61:0: W0311: Bad indentation. Found 17 spaces, expected 16 (bad-indentation)
```

> It must be fixed in another PR, but it goes out of the scope for this PR.

I'm fine with _fixing_ migrations in another PR, but we should set them to be ignored until then (and maybe just keep them ignored - I don't think it's reasonable to expect developers to fix up every problem in autogenerated code).

zhiltsov-max replied:
The list you mentioned should be covered by #5188.

The migrations are not that obvious. They are generated one time and then they live in the repo, they can be modified, and often they are. In the problematic case the code was modified and clearly has errors.

The strategy of not fixing everything and punishing the developers who touched the files that were not updated since then was accepted quite a long ago. This PR doesn't change anything in this regard. We can raise this topic and discuss it on the next retro.

SpecLad replied:
> The list you mentioned should be covered by https://github.com/opencv/cvat/pull/5188.

I think that covers all of them except for `utils/dicom_converter/script.py`.

> The migrations are not that obvious. They are generated one time and then they live in the repo, they can be modified, and often they are. In the problematic case the code was modified and clearly has errors.

Okay, fair enough.

> The strategy of not fixing everything and punishing the developers who touched the files that were not updated since then was accepted quite a long ago. This PR doesn't change anything in this regard. We can raise this topic and discuss it on the next retro.

Okay.

zhiltsov-max commented on lint-staged.config.js:
```diff
@@ -13,28 +13,49 @@ function makePattern(extension) {
     const cvatUI = containsInPath('/cvat-ui/', eslintFiles);
+    const cvatSdk = containsInPath('/cvat-sdk', pythonFiles);
+    const cvatCli = containsInPath('/cvat-cli', pythonFiles);
+    const pythonTests = containsInPath('/tests/python', pythonFiles);
```
Fixed

zhiltsov-max commented on lint-staged.config.js:
```diff
@@ -13,28 +13,49 @@ function makePattern(extension) {
+    const pythonTests = containsInPath('/tests/python', pythonFiles);
+
+    const tasks = []; // [command, file list] pairs
+    tasks.push(['npx stylelint --fix ', scssFiles.join(' ')]);
```
Fixed

zhiltsov-max commented:
Updated. I've changed the solution for Python, because the environment we are working in can vary significantly, and in some cases the hooks won't be available. The example of this is when we use the Git extension in VS Code and commit directly from it. The new solution allows Python tools to be missing in the environment, and it allows the user to set up a custom interpreter path, which is set to the place recommended by the installation instructions by default.

SpecLad commented:
> and it allows the user to set up a custom interpreter path

How? From what I can see, `.husky/pre-commit` always sets `PYTHON` to a fixed value, and then the actual checks use that value.

zhiltsov-max commented:
> How? From what I can see, .husky/pre-commit always sets PYTHON to a fixed value, and then the actual checks use that value.

The script can be modified after installation, it is only updated after `yarn prepare` is called. 

A possible option I see here is to create a separate env file, like VS Code uses. It's not clear if it is really needed.

SpecLad commented:
> The script can be modified after installation, it is only updated after yarn prepare is called.

That's not robust, because `yarn install` runs `yarn prepare`, and you need to re-run `yarn install` whenever the dependencies change.

I think you could add something like

```
if [ -r .husky/settings.sh ]; then
    . ./husky/settings.sh
fi
```

to `.husky/pre-commit`, and add `settings.sh` to `.gitignore`. That way, a developer can set up permanent variable overrides.

zhiltsov-max commented:
Ok, sounds good. I'll add this.

zhiltsov-max commented:
Closed till better times
